bin/propolis-server/src/lib/spec/api_spec_v0.rs

```diff
@@ -0,0 +1,259 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Conversions from version-0 instance specs in the [`propolis_api_types`]
+//! crate to the internal [`super::Spec`] representation.
+
+use propolis_api_types::instance_spec::{
+    components::devices::{SerialPort as SerialPortDesc, SerialPortNumber},
+    v0::{InstanceSpecV0, NetworkBackendV0, NetworkDeviceV0, StorageDeviceV0},
+};
+use thiserror::Error;
+
+#[cfg(feature = "falcon")]
+use propolis_api_types::instance_spec::components::devices::SoftNpuPort as SoftNpuPortSpec;
+
+#[cfg(feature = "falcon")]
+use crate::spec::SoftNpuPort;
+
+use super::{
+    builder::{SpecBuilder, SpecBuilderError},
+    Disk, Nic, QemuPvpanic, SerialPortUser, Spec,
+};
+
+#[derive(Debug, Error)]
+pub(crate) enum ApiSpecParseError {
+    #[error(transparent)]
+    BuilderError(#[from] SpecBuilderError),
+
+    #[error("backend {backend} not found for device {device}")]
+    BackendNotFound { backend: String, device: String },
+
+    #[error("backend {0} not used by any device")]
+    BackendNotUsed(String),
+
+    #[error("network backend for guest NIC {0} is not a viona backend")]
+    GuestNicInvalidBackend(String),
+
+    #[cfg(feature = "falcon")]
+    #[error("network backend for device {0} is not a DLPI backend")]
+    NotDlpiBackend(String),
```

Comment:
it might be nice if these indicated what the actual kind of backend we encountered was? but, not a big deal.

Reply:
I've left this alone for now because these variants are going to go away in a couple of PRs. (At least as I have things drafted the semantics of these variants will change from from "this NIC has a backend but its type is wrong" to "this NIC doesn't have a backend of the appropriate type".)